### DISCUSSION TITLE
Getting warning on automatic alias when migrating to 1.4.x

### DISCUSSION THREAD

**Discussion by fschulze** [sqlalchemy/sqlalchemy#8526] [Usage Questions] [answered]
I'm trying to migrate our application to SQLAlchemy 1.4.x. All the warnings so far have been super useful and pointed out actual problems/inefficiencies, so thanks for that!

I managed to get rid of almost all instances of ``sqlalchemy.exc.SAWarning: An alias is being generated automatically against joined entity mapped class ...``, but now encountered an issue where I'm stuck.

The problem is, that previously it was possible (though not necessarily working 100% according to the warning) to build upon existing queries and add additional filters to joined models. Now that I switched to use ``aliased``, the query can't be modified later on, because unlike the previously used model, the now used alias isn't accessible anymore as far as I can tell. Basically I return queries from properties or function calls and use them in joins or with additional filters. Is there a way to access the aliases and add filters to them when having only access to the query? It would be totally fine to do that by name, but in the current architecture of our application I can't pass the return value of ``aliased`` along. If that isn't possible, what is a better pattern for this? Would I have to convert properties to methods to be able to add additional filters via arguments?

Edit: I'm using ``flat=True`` as suggested, which disables the ability to add a name.

**Comment by fschulze** [comment, marked as answer]
Seems to be a classic case of rubber ducking:

Adding additional ``aliased`` when reusing the queries works correctly and the resulting queries are optimized to avoid unnecessary multiple joins.

**Reply by zzzeek** [reply]
I can follow your original question, for which I dont have great answers other than "maybe do it differently", or perhaps extract the alias object using [column_descriptions](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=column_descriptions#sqlalchemy.orm.Query.column_descriptions).      But your answer, I dont know what you're doing at all :)   feel free to illustrate a "before and after" example so I have something to use next time this question comes along.

**Comment by zzzeek** [comment]
this is a weak "law of twos" with #8495 :)

**Reply by CaselIT** [reply]
Now you are cheating! Let's cap the law time orizon to 1 week 😂

**Comment by fschulze** [comment]
More or less in pseudo code, as it would be too complex to provide a proper case.

Before 1.4.x:
```python

class ContentBase:
    ....

class Container(ContentBase):
    ....

class SubContainer(ContentBase):
    ....

class Item(ContentBase):
    ....

basequery = (
    session.query(Item)
    .join(SubContainer)
    .join(Container)
    .filter(
        Container.something == 'egg',
        SubContainer.created < sometime))

query = (
    basequery
    .filter(
        Container.another_attrib == 'ham'))
```

With 1.4.x I got the alias warnings. The ``basequery`` (which is actually a property) needed aliases for ``Container`` and ``SubContainer``. The problem was, that ``query`` then failed. I'm not sure anymore what exactly I got, it was either a warning about an implicit join, or about a missing join. What I figured out worked was to create a new alias for ``Container`` when reusing the ``basequery`` to build ``query`` and SQLAlchemy correctly added the new filter without wrapping everything into yet another join. Not sure how it figures all of that out, but it works and I'm glad it does. So: Thank you for SQLAlchemy, it is awesome!

**Comment by fschulze** [comment]
I saw ``column_descriptions``, but the filters aren't part of the result entity ``Item``.
